File: C#/Lists/LinkedSet/LinkedSet.cs
/*******************************************************
 *  Linkedset.cs
 *  Created by Stephen Hall on 11/20/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  Linked Set implementation in C#
 ********************************************************/
using System;


namespace DataStructures.Lists.LinkedSet
{    
    /// <summary>
    /// Linked set class
    /// </summary>
    /// <typeparam name="T">Generic type</typeparam>
    public class LinkedSet<T> where T : IComparable
    {
        /// <summary>
        /// Node class for singly linked set
        /// </summary>
        public class Node
        {
            /// <summary>
            /// Public accessors for node class
            /// </summary>
            public T Data { set; get; }
            public Node Next { set; get; }

            /// <summary>
            /// Node Class Constructor
            /// </summary>
            /// <param name="data">Data to be held in the Node</param>
            public Node(T data)
            {
                Data = data;
                Next = null;
            }
        }

        /// <summary>
        /// Private Members
        /// </summary>
        private Node _head;
        private Node _tail;
        private int _count;

        /// <summary>
        /// Linked Set Constructor
        /// </summary>
        public LinkedSet()
        {
            _head = _tail = null;
            _count = 0;
        }
        
        /// <summary>
        /// Adds a new node into the list with the given data
        /// </summary>
        /// <param name="data">Data to add into the list</param>
        /// <returns>Node added into the list</returns>
        public Node Add(T data)
        {
            // No data to insert into list
            if (data == null || (Find(data) != null))
                return null;

            Node node = new Node(data);
            // The Linked list is empty
            if (_head == null)
            {
                _head = node;
                _tail = _head;
                _count++;
            }
            else
            {
                // Add to the end of the list
                _tail.Next = node;
                _tail = node;
                _count++;
            }
            return node;
        }

        /// <summary>
        /// Removes the first node in the list matching the data
        /// </summary>
        /// <param name="data">Data to remove from the list</param>
        /// <returns>Node removed from the list</returns>
        public Node Remove(T data)
        {
            // List is empty or no data to remove
            if (_head == null || data == null)
                return null;

            Node tmp = _head;
            // The data to remove what found in the first Node in the list
            if (tmp.Data.Equals(data))
            {
                _head = _head.Next;
                _count--;
                return tmp;
            }
            // Try to find the node in the list
            while (tmp.Next != null)
            {
                // Node was found, Remove it from the list
                if (tmp.Next.Data.Equals(data))
                {
                    Node node = tmp.Next;
                    tmp.Next = tmp.Next.Next;
                    _count--;
                    return node;
                }
                tmp = tmp.Next;
            }
            // The data was not found in the list
            return null;
        }

        /// <summary>
        /// Gets the first node that has the given data
        /// </summary>
        /// <param name="data">Data to find in the list</param>
        /// <returns>First node with matching data or null if no node was found</returns>
        public Node Find(T data)
        {
            // No list or data to find
            if (_head == null || data == null)
                return null;

            Node tmp = _head;
            // Try to find the data in the list
            while (tmp != null)
            {
                // Data was found
                if (tmp.Data.Equals(data))
                    return tmp;
                tmp = tmp.Next;
            }
            // Data was not found in the list
            return null;
        }

        /// <summary>
        /// Gets the node at the given index
        /// </summary>
        /// <param name="index">Index of the Node to get</param>
        /// <returns>Node at passed in index</returns>
        public Node IndexAt(int index)
        {
            //Index was negative or larger then the amount of Nodes in the list
            if (index < 0 || index > Size())
                return null;

            Node tmp = _head;
            // Move to index
            for (int i = 0; i < index; i++)
                tmp = tmp.Next;
            // return the node at the index position
            return tmp;
        }

        /// <summary>
        /// Gets the current size of the list
        /// </summary>
        /// <returns>Number of items in the list</returns>
        public int Size() => _count;
    }
}

File: PHP/Lists/LinkedSet/LinkedSet.php
<?php namespace datastructures\linkedset;
/*******************************************************
 *  Linkedset.php
 *  Created by Stephen Hall on 04/24/18.
 *  Copyright (c) 2018 Stephen Hall. All rights reserved.
 *  Linked Set implementation in PHP
 ********************************************************/

/**
 * Node class for singly linked set
 */
class Node{
    /**
    * Public Members
    */
    public $data = null;
    public $next = null;

    /**
    * Node Class Constructor
    * @param $data : Data to be held in the Node
    */
    public function __construct($data){
        $this->data = $data;
        $this->next = null;
    }
}

/**
 * Linked set class
 */
class LinkedSet {

    /**
     * Private Members
     */
    private $head;
    private $tail;
    private $count;

    /**
     * Linked Set Constructor
     */
    public function __construct(){
        $this->head = $this->tail = null;
        $this->count = 0;
    }

    /**
     * Adds a new node into the list with the given data
     * @param $data : Data to add into the list<
     * @return Node added into the list
     */
    public function add($data){
        // No data to insert into list
        if ($data == null || ($this->find($data) != null))
            return null;

        $node = new Node($data);
        // The Linked list is empty
        if ($this->head == null) {
            $this->head = $node;
            $this->tail = $this->head;
            $this->count++;
        } else {
            // Add to the end of the list
            $this->tail->next = $node;
            $this->tail = $node;
            $this->count++;
        }
        return $node;
    }

    /**
     * Removes the first node in the list matching the data
     * @param $data : Data to remove from the list
     * @return Node removed from the list
     */
    public function remove($data){
        // List is empty or no data to remove
        if ($this->head == null || $data == null)
            return null;

        $tmp = $this->head;
        // The data to remove what found in the first Node in the list
        if($this->equalTo($tmp->data, $data)) {
            $this->head = $this->head->next;
            $this->count--;
            return $tmp;
        }
        // Try to find the node in the list
        while ($tmp->next != null) {
            // Node was found, Remove it from the list
            if ($this->equalTo($tmp->next->data, $data)){
                $this->node = $tmp->next;
                $tmp->next = $tmp->next->next;
                $this->count--;
                return $tmp;
            }
            $tmp = $tmp->next;
        }
        // The data was not found in the list
        return null;
    }

    /**
     * Gets the first node that has the given data
     * @param $data : Data to find in the list
     * @return Node First node with matching data or null if no node was found
     */
    public function find($data){
        // No list or data to find
        if ($this->head == null || $data == null)
            return null;

        $tmp = $this->head;
        // Try to find the data in the list
        while($tmp != null) {
            // Data was found
            if ($this->equalTo($tmp->data, $data))
                return $tmp;
            $tmp = $tmp->next;
        }
        // Data was not found in the list
        return null;
    }

    /**
     * Gets the node at the given index
     * @param $index : Index of the Node to get
     * @return Node at passed in index
     */
    public function indexAt($index){
        //Index was negative or larger then the amount of Nodes in the list
        if ($index < 0 || $index > $this->size())
            return null;

        $tmp = $this->head;
        // Move to index
        for ($i = 0; $i < $index; $i++)
            $tmp = $tmp->next;
        // return the node at the index position
        return $tmp;
    }

    /**
     * Gets the current count of the array
     * @return Number of items in the array
     */
    public function size(){
        return $this->count;
    }

    /**
     * Determines if a is equal to b
     * @param $a: generic type to test
     * @param $b: generic type to test
     * @return boolean: true|false
     */
    private function equalTo($a, $b) {
        return $a === $b;
    }
}